state/block-executor.md

```diff
@@ -0,0 +1,124 @@
+
+- `InitChain`: This method initializes the chain by calling ABCI `InitChainSync` using the consensus connection to the app. It takes a `GenesisDoc` as a parameter. It sends a ABCI `RequestInitChain` message with the genesis parameters including:
+  - Genesis Time
+  - Chain ID
+  - Consensus Parameters including:
+    - Block Max Bytes
+    - Block Max Gas
+    - Evidence Parameters
+    - Validator Parameters
+    - Version Parameters
+  - Initial Validator Set using genesis validators
+  - Initial Height
+
+- `CreateBlock`: This method reaps transactions from the mempool and builds a block. It takes the state, the height of the block, last header hash, and the commit as parameters.
+
+- `ApplyBlock`: This method applies the block to the state. Given the current state and block to be applied, it:
+  - Validates the block, as described in `Validate`.
+  - Executes the block using app, as described in `execute`.
+  - Captures the validator updates done in the execute block.
+  - Updates the state using the block, block execution responses, and validator updates as described in `updateState`.
+  - Returns the updated state, validator updates and errors, if any, after applying the block.
+  - It can return the following named errors:
+
+    - `ErrEmptyValSetGenerate`: returned when applying the validator changes would result in empty set.
+    - `ErrAddingValidatorToBased`: returned when adding validators to empty validator set.
+
+- `Validate`: This method validates the block. It takes the state and the block as parameters. In addition to the basic [block validation] rules, it applies the following validations:
+
+  - New block version must match block version of the state.
+  - If state is at genesis, new block height must match initial height of the state.
+  - New block height must be last block height + 1 of the state.
+  - New block header `AppHash` must match state `AppHash`.
+  - New block header `LastResultsHash` must match state `LastResultsHash`.
+  - New block header `AggregatorsHash` must match state `Validators.Hash()`.
+
+- `updateState`: This method updates the local state using the block and block execution responses and returns the updated state and errors, if any. Given the current state, the block and the ABCI `ResponseFinalizeBlock`, it validates the updated validator set and updates the state by applying the block. The state consists of:
+  - Version
+  - Chain ID
+  - Initial Height
+  - Last Block including:
+    - Block Height
+    - Block Time
+    - Block ID
+  - Next Validator Set
+  - Current Validator Set
+  - Last Validators
+  - Whether Last Height Validators changed
+  - Consensus Parameters
+  - Whether Last Height Consensus Parameters changed
+  - App Hash
+
+- `commit`: This method commits the block and updates the mempool. Given takes the updated state, the block, and the ABCI `ResponseFinalizeBlock` as parameters, it:
```

Comment:
* `Commit` is public function in 0.11.x
* `Given takes` probably typo?

Reply:
Fixed.